src/adaptation/icap/Xaction.cc

```diff
@@ -187,12 +188,17 @@ Adaptation::Icap::Xaction::openConnection()
     debugs(93,3, typeName << " opens connection to " << s.cfg().host.termedBuf() << ":" << s.cfg().port);
 
     // Locate the Service IP(s) to open
+    assert(!waitingForDns);
```

Comment:
Can we use Must() here instead of assert() ? to only affect the transactions made within the window of brokenness. Other transactions once the data is available should not be prevented by Squid having shutdown.

Reply:
> It would be good to only affect the transactions made within the window of brokenness

Yes, we are on the same page here.

> Can we use Must() instead of assert() for these checks?

Yes, we can, but I do not think we should. I have finally completed the corresponding Error Handling nodes that argue [against](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-exceptions-vs-assertions) long-term assert() use and [against](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api) Must()s in new code like this.

Let's use this change request to decide how to check invariants here and/or in future PRs. I propose adding Assert() and can volunteer to do that (in a separate PR) if we decide that we should go in that direction (this PR code can wait or be merged as is). I cannot volunteer to check and refactor 900 Must()s, unfortunately. See the notes linked above for the details.

Reply:
I think the discussion around Must() and any missing exception APIs is an integral part of the design discussion we are going to have about long term try-catch placements. As such better in squid-dev@ than any github PR.

Since we are on the same page about how bad assert is for this and similar. Can we at least throw an exception if you disagree with Must() use as a "best we have" API? this seems to be a rather serious bug to block on designing perfect error handling in Squid.

Reply:
> Can we at least throw an exception if you disagree with Must() use as a "best we have" API

In this context: 

* I believe assert() is the best short-term solution.
* Must() is worse than assert() because it hides the bug, as now unofficially documented [elsewhere](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api).
* Explicitly throwing an exception is worse than Must() because in addition to hiding the bug like Must() would do, it misleads the reader into thinking that this is input validation (rather than a code logic/invariant check). This option is also needlessly verbose, duplicating Must() functionality 1:1.
* Combining manual exception throwing with a custom level-0 debugs("BUG") statement (to expose the bug) is worse than assert() because it adds and then duplicates custom code that will need to be found and replaced with Assert() or similar. If we go down this route, then it is much better to just add and use Assert() instead!
* I am not against adding and using throwing-and-logging Assert(). It the best long-term solution AFAICT. It can be done in a dedicated prerequisite PR.

> I think the discussion around Must() and any missing exception APIs is an integral part of the design discussion we are going to have about long term try-catch placements. As such better in squid-dev@ than any github PR.

Your call.

> this seems to be a rather serious bug to block on designing perfect error handling in Squid.

I agree with the implication that this PR should not be blocked. Please consider unblocking. AFAICT, this PR code follows one of the well-known, wide-spread, valid approaches that modern Squid code uses for invariant checks. That approach is not perfect, but if we cannot agree on a replacement, then I see no valid reasons to block the PR. Needless to say, I am not blocking it.

Reply:
Its clear this will need an XXX to highlight the bug then.

Reply:
Which bug?! I have no idea what XXX/problem you are referring to. You are not implying that all suspected transaction-scope asserts() will now need `XXX: Treat as a transaction-scope bug` comments, do you? If that is indeed the implication, then, again, it is much better to add and use Assert() (or similar) instead of sprinkling the code with such comments.

Reply:
My current understanding is that this assert() is possible to trigger from a (rare) combination of server and client input timings. Which makes it a bug that they can trigger a full crash and DoS of Squid.

Reply:
> My current understanding is that this assert() is possible to trigger from a (rare) combination of server and client input timings.

What makes you think that? My current understanding is that there is no (known or suspected) way to trigger this assert().

Reply:
The third paragraph of the description:
> However, some REQMOD and RESPMOD transactions are still affected by this
> bug because they require an OPTIONS transaction if the service options
> have never been fetched before or have expired. For example, the first
> few transactions for a given service -- all those started before the DNS
> lookup completes and Squid caches its result -- will fail.

Those "will fail" transactions happen between DNS being requested and found. Thus service's waitingForDns is true, making the assert happen.

If no, what am I misreading?

Reply:
I rewrote the PR description in hope to correct this misunderstanding. I am sorry that my original description was difficult to interpret correctly. I hope the current version is clear about what was wrong with REQMOD/RESPMOD transactions and cannot be easily misinterpreted as if they are still suffering from the fixed bug. Please check.

Reply:
Oh, I see. Confusion was the tense of the word "could" differs between our variants of english. FYI, in my locale it is typically used as present-tense instead of "can" or "might" rather than past-tense.